Security Response Header Uniformity Improvements

jentfoo opened the pull request:
When investigating how to resolve #12529 several discoveries were made:

1) We were in some cases already applying the Strict Transport response header: https://github.com/gravitational/teleport/blob/master/lib/httplib/httpheaders.go#L59

2) We had logic for redirects which deferred to the above function, and then replaced the Content-Security-Policy and the Referrer-Policy: https://github.com/gravitational/teleport/blob/master/lib/web/app/redirect.go#L28

3) Neither of which were uniformly applied across the service, resulting in the initially mentioned issue

Our need for customizing our security response headers primarily exists within the Content-Security-Policy.  Other headers can be applied more broadly to provide safe defaults across our service.  For that reason this PR attempts to centralize the logic for these "default" security headers.

At a high level what I have done here is:
* Change `httpheaders.go` to allow for the CSP to be applied separate from the other headers.  Then the changes to `httplib.go` 's handlers as the primary method of making sure these defaults are applied in a common way (similar to how we were already doing for caching directives).
* Change the redirect logic to better fit this model and fix a few places where redirects were returned but without the redirect header logic being applied.
* Simplified several cases where this logic was being handled as a one off but is now being handled centrally

I have added some testing to `apiserver_test.go` but I believe we likely will want this verification in more places.

jentfoo committed:
Some minor initial cleanup before making larger changes

jentfoo committed:
Centralize security response header logic

Our need for customizing our security response headers primarily exists with the Content-Security-Policy.  Other headers can be applied more broadly to provide safe defaults across our service.

This commit changes httpheaders.go to allow for the CSP to be applied separate from the other headers.  Then the changes to httplib.go's handlers help make sure these headers are set generally (which we were already partially doing for caching directives).

Other changes were primarily making sure these tools are configured or used.  Or a removal of previous settings as this function is now being handled more centrally.

zmb3 reviewed (commented):
No major concerns with the approach, but I'm still having a hard time verifying whether the behavior is identical after this change.

jentfoo commented on lib/web/apiserver_test.go:
```diff
@@ -2422,6 +2466,7 @@ func TestInstallDatabaseScriptGeneration(t *testing.T) {
 			},
 		})
 	require.NoError(t, err)
+	verifySecurityResponseHeaders(t, re.Headers())
```
I considered something similar.  But it seemed too complicated to try and exercise the entire API space in a single test (we would basically need to go through the sum of all the actions being done here).  That's why I was instead trying to put verification around existing responses we conveniently already had.

We could create a single test and only test one or two api endpoints, but I was hoping to get more verification coverage here (particularly since I am trying to verify my centralized fixes are functioning broadly)

I also considered we could have the tests save all the results and then we can have a single test execute at the end and review / verify them.  But this would introduce test ordering requirements :/

All in all, I don't really like how this testing looks, and I know it's not complete.  Some non-ideal solution will probably need to be accepted, so let me know you think might be best.

jentfoo replied:
I was able to detect and fix the issue mentioned yesterday by spreading this testing pattern further through this test class.  @zmb3 I would appreciate your perspective on if I should:
1) Leave this check spread throughout the test file
2) Keep this pattern but minimize it to just a handful of points
3) Adopt one of the strategies above

I personally lean towards option 1 because it's most complete with minimal code, but I am happy with any solution.

jentfoo commented:
@zmb3 helped demonstrate a behavior difference I did not expect.  This PR is not currently functioning correctly as the caching headers were not being applied as broadly as I thought.  In testing that shows the security headers also having gaps in their functionality.

I will let everyone know once I get to the bottom of this

jentfoo committed:
Fix apiserver header behavior

This expands the tests in apiserver_test.go as a means to help debug and demonstrate the header behavior change.
The testing showed the need to unconditionally set the security headers at the start of the http handler in apiserver.go.

jentfoo commented:
With the most recent commit I believe the service behavior is matching what we expect.  I did some manual testing of my local instance vs cloud and I am not seeing any response header differences that I would not expect from this change.

I think the testing is still not complete or optimal.  I welcome feedback, thank you!

jentfoo committed:
Document SetDefaultSecurityHeaders and remove extra h.Set call

nklaassen reviewed (approved):
It all looks good to me. I left a comment on the testing but I think what you have is alright if you just want to go ahead with it.

jentfoo committed:
middleware.go: Make tracing handler the outer-most wrapped handler

Request from PR to help ensure visibility from the tracing handler is done first.

jentfoo committed:
apiserver_test.go - Move security header verification to the test client

jentfoo commented:
@zmb3 and @nklaassen In the most recent commit I followed the suggestion to put the header verification in the test client.  I go back and forth on what's best, so just let me know if you prefer it with that change or if I should revert it.

Thank you!

nklaassen commented on lib/web/apiserver_test.go:
```diff
@@ -5946,16 +5901,30 @@ func waitForOutput(stream *terminalStream, substr string) error {
-func (s *WebSuite) login(clt *client.WebClient, cookieToken string, reqToken string, reqData interface{}) (*roundtrip.Response, error) {
+type TestWebClient struct {
+	*client.WebClient
+	t *testing.T
+}
+
+func (c *TestWebClient) RoundTrip(fn roundtrip.RoundTripFn) (*roundtrip.Response, error) {
+	c.t.Helper()
+	resp, err := c.WebClient.RoundTrip(fn)
+
+	verifySecurityResponseHeaders(c.t, resp.Headers())
```
The potential issue with holding onto `t` like this is that if the client is passed to a subtest, or used on a goroutine other than the one running the test, it will not be correct to call `t.FailNow()` (which is what the `testify/require` functions will do). It may be enough to just have `RoundTrip` return an error if the security headers are not there.

nklaassen replied:
Other than that, I think I do like having the check in the client.

nklaassen replied:
I was thinking about this and just realized, the other thing to note is that since Go does not do dynamic dispatch, your `(*TestWebClient).RoundTrip` will only run if called directly - if the test just calls `clt.Get` or something else, that will go to the embedded `*client.WebClient` and it will not call back out to anything from `TestWebClient`.

If we think it's worth running this on all client calls, I think you'd either have to wrap each `WebClient` method, or install a custom `http.RoundTripper` to the underlying `http.Client.Transport`

jentfoo replied:
Sorry, I should have commented that I was aware of this.  I removed my changes in the service and made sure that this minimal verification was able to verify my changes (ie the tests fail).  I tried several permutations and felt like that even though this was not absolute it was possibly "good enough" while being simple.

If we do want to extend to the other functions I felt like the simplicity was sort of loss and maybe the former implementation was better.  But again, I don't feel strongly one way or another, but it was recognized.

nklaassen replied:
Yeah I think it should be good enough, it just broke my original assumption when I realized that so thought I'd comment in case it was missed, glad to know it was considered.

jentfoo replied:
I left a comment to make sure there is no confusion and also committed a simplification that was missed earlier.  Thank you for raising it!

jentfoo committed:
apiserver_test.go - Clarifications

Added a comment about where RoundTrip logic is expected to be run, and also simplified where we are constructing the TestWebClient struct.
